appsec/src/extension/tags.c

```diff
@@ -1095,85 +1270,150 @@ static PHP_FUNCTION(datadog_appsec_track_user_login_failure_event)
         return;
     }
 
-    zend_string *user_id = NULL;
-    zend_bool exists = false;
+    zend_string *user_login;
+    zend_string *user_id;
+    zend_string *anon_user_login = NULL;
+    zend_string *anon_user_id = NULL;
+    zend_bool exists;
     HashTable *metadata = NULL;
-    zend_bool automated = false; // Don't document. Only internal usage
-    if (zend_parse_parameters(ZEND_NUM_ARGS(), "Sb|hb", &user_id, &exists,
-            &metadata, &automated) == FAILURE) {
+    if (zend_parse_parameters(ZEND_NUM_ARGS(), "SSb|h", &user_login, &user_id,
+            &exists, &metadata) == FAILURE) {
         mlog(dd_log_warning, "Unexpected parameter combination, expected "
-                             "(user_id, exists, metadata)");
+                             "(user_login, user_id, exists, metadata)");
+        return;
+    }
+
+    zval *nullable meta = _root_span_get_meta();
+    if (!meta) {
         return;
     }
 
-    zend_bool copy_user_id = true;
-    if (automated) {
-        user_collection_mode mode = dd_get_user_collection_mode();
-        if (mode == user_mode_disabled ||
-            !get_DD_APPSEC_AUTOMATED_USER_EVENTS_TRACKING_ENABLED()) {
+    _user_event_triggered = true;
+    zend_array *meta_ht = Z_ARRVAL_P(meta);
+
+    user_collection_mode mode = dd_get_user_collection_mode();
+    if (mode == user_mode_disabled ||
+        !get_DD_APPSEC_AUTOMATED_USER_EVENTS_TRACKING_ENABLED()) {
+        return;
+    }
+
+    if (mode == user_mode_anon) {
+        anon_user_id = dd_user_info_anonymize(user_id);
+        if (!anon_user_id) {
+            mlog(dd_log_debug, "Failed to anonymize user ID");
             return;
         }
 
-        if (mode == user_mode_anon) {
-            user_id = dd_user_id_anonymize(user_id);
-            if (user_id == NULL) {
-                mlog(dd_log_debug, "Failed to anonymize user ID");
-                return;
-            }
-
-            copy_user_id = false;
+        anon_user_login = dd_user_info_anonymize(user_login);
+        if (!anon_user_login) {
+            mlog(dd_log_debug, "Failed to anonymize user login");
+            zend_string_release(anon_user_id);
+            return;
+        }
 
-            if (metadata != NULL && zend_array_count(metadata) > 0) {
-                metadata = NULL;
-            }
+        if (metadata && zend_array_count(metadata) > 0) {
+            metadata = NULL;
         }
     }
 
+    if (ZSTR_LEN(user_id) > 0) {
```

Comment:
This looks like it's leaking `anon_user_id` / `anon_user_login` if `user_id` has length 0 (and `mode == user_mode_anon`)

Reply:
Fixed it and added some tests to make sure it's fine